### PR TITLE
refactor: port cucumber tests to rs

### LINKED ISSUE #5050: Convert Cucumber tests
# Problem

We are using a javascript framework for testing. We can now move to a rust based framework and believe this will be easier to manage in the longrun. We need to port the existing cucumber tests over, and reimplement all the steps in rust.

The problem is less the need for the port but in the execution itself. It's a lot of code to move. More people involved can create conflicts and unneeded churn. A strategy to move quickly without stepping on toes is important.

# Appetite
The appetite for minimum 80% completion of this port is 2.5 weeks. Ending on December 30th. Failure to meet completion may result in the abandonment of this port. Check-ins:
- [x] Friday 16th: Is the ground work feature branch ready
- [x] Tuesday 20th: What percentage of steps have been ported, 16% needed to stay on minimum target
- [x] Thursday 22nd: What percentage of steps have been ported, 36% needed to stay on minimum target
- [ ] 2nd week TBD

# Solution
It's easy to port the .feature files into the rust library. The difficulty is enabling each of the needed steps without duplicating work. I propose we spend time implementing steps in isolation and uncommenting implemented steps in all tests that use the step assuming this will make that test pass and reduce the chance of conflict. 

Some quick numbers:
- 18% of steps make up 80% of all step invocations
- 29% of steps that are completely unique (used once)
- 15% of steps aren’t used at all, and require no porting 🎉

Implementing 18% of the steps in isolation gives us much more breathing room to begin porting/uncommenting entire tests in quicker succession and with less conflict.

We already have a [base feature branch](feature-cucumber-rs) to work from.

The process for this work will be:
- to check out a branch from the provided base branch
- Assign yourself a single/or set of steps you will implement (see spreadsheet not attached)
- Write the implementations
- Uncomment the step in tests, and ensure it passes
- Create a PR back into the base feature branch

These PR's can be made in rapid succession, and contain as little or as much as you've completed. Had 10 minutes to implement a single step, send the PR. Finished 8 steps but working on a 9th and it's the end of day, submit the PR with 8. We're looking for speed here.

# Rabbit Holes
- If we can eliminate unique or single-use steps in favor of an existing step it can be worth the exercise, but don't devote much time to the effort.

# No Gos
- No writing net new features, we're not looking to increase coverage at the moment.

### PR DESCRIPTION
**Pull request by brianp**
Description
---
This is a Base feature branch that all additional work on the implementation will be merged into, before being merged into development.

Motivation and Context
---
Removing the JS dependencies in favor of a pure rust implementation

Related
---
Toolchain upgrade:
https://github.com/tari-labs/docker/pull/23
https://github.com/tari-project/action-buildlibs/pull/10

How Has This Been Tested?
---
CI

Closes: #5050

### CODE REVIEW

File: base_layer/core/src/base_node/comms_interface/comms_response.rs [hunk 1]
```diff
@@ -38,6 +38,7 @@ use crate::{
 };
 
 /// API Response enum
+#[allow(clippy::large_enum_variant)]
```
Comment (SWvheerden) [inline base_layer/core/src/base_node/comms_interface/comms_response.rs:41]: Why is this now needed?
It does not look like we changed the enum

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs [hunk 2]
```diff
@@ -265,7 +265,7 @@ impl UnconfirmedPool {
         transactions_to_recheck: &mut Vec<(TransactionKey, Arc<Transaction>)>,
         selected_txs: &HashMap<TransactionKey, Arc<Transaction>>,
         total_weight: &mut u64,
-        unique_ids: &mut HashSet<[u8; 32]>,
+        _unique_ids: &mut HashSet<[u8; 32]>,
```
Comment (SWvheerden) [inline base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs:268]: why this change?

**Review by SWvheerden — COMMENTED** [review]
The look good. 
I have not reviewed the cucumber tests only the changes in the core code.

File: base_layer/core/src/base_node/comms_interface/comms_response.rs (hunk 1 above)
Comment (brianp) [reply base_layer/core/src/base_node/comms_interface/comms_response.rs:41]: The toolchain changed, and now clippy is complaining about different/new things.

File: base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs (hunk 2 above)
Comment (brianp) [reply base_layer/core/src/mempool/unconfirmed_pool/unconfirmed_pool.rs:268]: Same here, was a clippy complaint on the new toolchain.

**Review by stringhandler — APPROVED** [review]

File: base_layer/core/src/consensus/emission.rs [hunk 3]
```diff
@@ -136,7 +136,7 @@ impl EmissionSchedule {
             let mut carry_last = 0u8;
             for i in 0..len {
                 let index = len - 1 - i;
-                let carry = if num[index] >= 5 { 1 } else { 0 };
+                let carry = (num[index] >= 5).into();
```
Comment (stringhandler) [inline base_layer/core/src/consensus/emission.rs:139]: Not a fan of this. I think a clippy allow would be clearer

File: base_layer/core/src/proof_of_work/monero_rx/merkle_tree.rs [hunk 4]
```diff
@@ -224,8 +224,8 @@ pub fn create_merkle_proof(hashes: &[Hash], hash: &Hash) -> Option<MerkleProof>
             if h != hash {
                 return None;
             }
-            let i = if pos == 0 { 1 } else { 0 };
-            MerkleProof::try_construct(vec![hashes[i]], 1, if pos == 0 { 0 } else { 1 })
+            let i = usize::from(pos == 0);
```
Comment (stringhandler) [inline base_layer/core/src/proof_of_work/monero_rx/merkle_tree.rs:227]: Also not a fan here
